File: 6_kyu/Duplicate_Encoder.rs
use std::collections::HashMap;

// Returns a string with the character '(' to indicate a single character in the word, and ')' to indicate characters used twice.
// Function ignores upper/lower case. (Uses lower case transformation).
fn duplicate_encode(word:&str)->String {
    // Converts all to the same case.
    let word = String::from(word).to_lowercase();
    // Create mapping to the indication character.
    let mut mapping = HashMap::<char,char>::with_capacity(word.len());
    for c in word.chars() {
        if let Some(item) = mapping.get_mut(&c) {
            *item = ')';
        } else {
            mapping.insert(c, '(');
        }
    }
    // Changes all chars to the two indication chars. 
    word.chars().map(|c| mapping.get(&c).unwrap()).collect()
}

File: 6_kyu/Duplicate_Encoder.r
# Returns a string with the char '(' to indicate a single char in the word,
# and ')' to indicate chars used twice or more in the word.
# Ignores upper/lower case.
duplicate_encode <- function(word) {
  word <- tolower(word)
  counts <- table(strsplit(word,"",fixed=TRUE))
  once <-paste_all( names(counts)[counts==1] )
  some <- paste_all( names(counts)[counts>1] )
  chartr(paste_all(once,some),paste_all(strrep("(",nchar(once)),strrep(")",nchar(some))),word)
}

# Pasts all arguments into one string without separators.
paste_all <- function(...) paste0(c(...),collapse="")
